tests/src/test/java/io/streamnative/pulsar/handlers/kop/TransactionTest.java

```diff
@@ -517,4 +520,64 @@ private String getKafkaServerAdder() {
     protected void addCustomizeProps(Properties producerProps) {
         // No-op
     }
+
+    @DataProvider(name = "isolationProvider")
+    protected Object[][] isolationProvider() {
+        return new Object[][]{
+                {"read_committed"},
+                {"read_uncommitted"},
+        };
+    }
+
+    @Test(dataProvider = "isolationProvider", timeOut = 1000 * 30)
+    public void readUnStableMessagesTest(String isolation) throws InterruptedException, ExecutionException {
+        String topic = "unstable-message-test-" + RandomStringUtils.randomAlphabetic(5);
+
+        KafkaConsumer<Integer, String> consumer = buildTransactionConsumer("unstable-read", isolation);
+        consumer.subscribe(Collections.singleton(topic));
+
+        String tnxId = "txn-" + RandomStringUtils.randomAlphabetic(5);
+        KafkaProducer<Integer, String> producer = buildTransactionProducer(tnxId);
+        producer.initTransactions();
+
+        String baseMsg = "test msg commit - ";
+        producer.beginTransaction();
+        producer.send(new ProducerRecord<>(topic, baseMsg + 0)).get();
+        producer.send(new ProducerRecord<>(topic, baseMsg + 1)).get();
+        producer.flush();
+
+        AtomicInteger messageCount = new AtomicInteger(0);
+        // make sure consumer can't receive unstable messages in `read_committed` mode
+        readAndCheckMessages(consumer, baseMsg, messageCount, isolation.equals("read_committed") ? 0 : 2);
+
+        producer.commitTransaction();
+        producer.beginTransaction();
+        // these two unstable message shouldn't be received in `read_committed` mode
+        producer.send(new ProducerRecord<>(topic, baseMsg + 2)).get();
+        producer.send(new ProducerRecord<>(topic, baseMsg + 3)).get();
+        producer.flush();
+
+        readAndCheckMessages(consumer, baseMsg, messageCount, isolation.equals("read_committed") ? 2 : 4);
+        ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofSeconds(3));
+        assertTrue(records.isEmpty());
+
+        consumer.close();
+        producer.close();
+    }
+
+    private void readAndCheckMessages(KafkaConsumer<Integer, String> consumer, String baseMsg,
+                                      AtomicInteger messageCount, int expectedMessageCount) {
+        while (true) {
+            ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofSeconds(3));
+            if (records.isEmpty()) {
```

Comment:
Should we retry several times when the records are empty?

Reply:
Maybe we can add a check after this poll loop.

Reply:
Yes, LGTM.